open-craft/opencraft
# [BB-4141] Add task for daily email report

Adds the daily report task to `marketing` app

### Related Tickets
[BB-4141](https://tasks.opencraft.com/browse/BB-4141)

### Testing Instructions

#### Testing report email is sent after sending followup emails.
1. Checkout to this branch
2. Run `make migrate` if you haven't already after the addition of marketing app.
4. Run dev server
5. Visit http://localhost:5000/admin/marketing/emailtemplate/add/ and add an active template with send_after_days set to 1.
6. Visit http://localhost:5000/admin/marketing/emailtemplate/add/ and add a subscriber with `trial_started_at` set to yesterday's date and a time less than now.
7. Feel free to create more records of `EmailTemplate` and `Subscriber` with different values
5. In another terminal open the django shell using `make shell`
6. Import the tasks using `from marketing.tasks import send_followup_emails, send_report`
7. Run `send_followup_emails()`
8. In server logs, verifiy that the report was sent for all the triggered emails.
9. Run the `send_followup_emails()` task again and verify that this time the report email says `No emails were sent today to P&T and trail users.`

#### Testing `send_report` standalone

1. In django shell import `send_report` task
2. `send_report` accepts a dictionary with `Email names` as key and `List of emails` as values.
3. Trigger `send_report` task using `send_report({"Welcome Email": ["arjun@example.com", "user2@example.com"]})`
4. In server logs, verify that the email is sent with appropriate template.
5. Change the values (increase keys and emails) and call `send_report` task.
6. Verify that email is sent with all the records passed.

### Author Notes

1. The `send_report` task is scheduled right after the `send_followup_email` with a list of followup emails send. I think it made more sense there because we wouldn't have to make a DB query to get the sent emails and it is the next step right after sending all followup emails.
